File: src/main/java/com/conference/website/api/dto/DtoConversions.java
package com.conference.website.api.dto;

import com.conference.website.domain.*;
import jakarta.validation.constraints.NotNull;
import org.jspecify.annotations.Nullable;

import java.util.Comparator;
import java.util.List;

public final class DtoConversions {

    private DtoConversions() {
    }

    public static SpeakerDto toDto(@NotNull Speaker speaker) {
        return new SpeakerDto(
                speaker.getId(),
                speaker.getName(),
                speaker.getEmail(),
                speaker.getCompany(),
                speaker.getBio()
                );
    }

    public static TagDto toDto(Tag tag) {
        return new TagDto(tag.getId(), tag.getName());
    }

    public static @Nullable ScheduleSlotDto toScheduleSlotResponse(@Nullable ScheduleSlot slot) {
        if (slot == null) {
            return null;
        }

        return new ScheduleSlotDto(
                slot.getId(),
                slot.getRoomName(),
                slot.getStartTime(),
                slot.getEndTime()
        );
    }

    public static TalkDto toDto(Talk talk) {
        List<RatingDto> ratingDto = talk.getRatings().stream()
                .sorted(Comparator.comparing(Rating::getCreatedAt).reversed())
                .map(DtoConversions::toDto)
                .toList();

        double averageRating = talk.getRatings().stream()
                .mapToInt(Rating::getScore)
                .average()
                .orElse(0.0d);

        return new TalkDto(
                talk.getId(),
                talk.getTitle(),
                talk.getAbstractText(),
                talk.getLevel(),
                talk.getDurationMinutes(),
//                talk.getCreatedAt(),
                toDto(talk.getPrimarySpeaker()),
                talk.getCoSpeakers().stream().map(DtoConversions::toDto).toList(),
                talk.getTags().stream().map(DtoConversions::toDto).toList(),
                ratingDto,
                toScheduleSlotResponse(talk.getScheduleSlot()),
                averageRating,
                (long) talk.getRatings().size()
        );
    }

    private static RatingDto toDto(Rating rating) {
        return new RatingDto(
                rating.getId(),
                rating.getReviewerName(),
                rating.getScore(),
                rating.getComment()
                //rating.getCreatedAt()
        );
    }

}

File: src/test/kotlin/com/conference/website/utils/DtoConversions.kt
package com.conference.website.utils

import com.conference.website.domain.Rating
import com.conference.website.domain.ScheduleSlot
import com.conference.website.domain.Speaker
import com.conference.website.domain.Tag
import com.conference.website.domain.Talk
import com.conference.website.dto.CreateSpeakerRequest
import com.conference.website.dto.CreateTalkRequest
import com.conference.website.dto.RatingDto
import com.conference.website.dto.ScheduleSlotDto
import com.conference.website.dto.ScheduleSlotRequest
import com.conference.website.dto.SpeakerDto
import com.conference.website.dto.TagDto
import com.conference.website.dto.TalkDto
import com.conference.website.dto.TestDtoConversions

fun CreateSpeakerRequest.toDto(id: Long): SpeakerDto = TestDtoConversions.toDto(id, this)

fun Speaker.toDto(): SpeakerDto = TestDtoConversions.toDto(this)

fun Tag.toDto(): TagDto = TestDtoConversions.toDto(this)

fun Rating.toDto(): RatingDto = TestDtoConversions.toDto(this)

fun ScheduleSlotRequest.toDto(id: Long): ScheduleSlotDto = TestDtoConversions.toDto(id, this)

fun ScheduleSlot.toDto(): ScheduleSlotDto = TestDtoConversions.toDto(this)

fun Talk.toDto(): TalkDto = TestDtoConversions.toDto(this)

fun CreateTalkRequest.toDto(
    id: Long,
    speaker: SpeakerDto,
    coSpeakers: List<SpeakerDto>,
    tagIds: List<TagDto>,
    ratings: List<RatingDto>,
    scheduleSlotDto: ScheduleSlotDto,
): TalkDto = TestDtoConversions.toDto(id, speaker, coSpeakers, tagIds, ratings, scheduleSlotDto, this)

fun CreateTalkRequest.toDto(
    id: Long,
    speaker: SpeakerDto,
    scheduleSlotDto: ScheduleSlotDto,
): TalkDto = TestDtoConversions.toDto(id, speaker, scheduleSlotDto, this)

fun CreateTalkRequest.toDto(id: Long): TalkDto = TestDtoConversions.toDto(id, this)

